### PR TITLE
Fix segfault in toLvalueImpl

### PR DESCRIPTION
**Pull request by dkorpel**
@ibuclaw I think this fixes the regression

### CODE REVIEW

File: compiler/src/dmd/expressionsem.d [hunk 1]
```diff
@@ -15218,7 +15218,9 @@ extern(C++) Expression toLvalue(Expression _this, Scope* sc)
 }
 
 // e = original un-lowered expression for error messages, in case of recursive calls
-private Expression toLvalueImpl(Expression _this, Scope* sc, Expression e) {
+private Expression toLvalueImpl(Expression _this, Scope* sc, Expression e)
+{
+    assert(e);
```
Comment (maxhaton) [inline compiler/src/dmd/expressionsem.d:15223]:
```suggestion
{
    assert(e);
    Expression makeANameUp = e ? e : _this.e1;
    // or 
    e = e ? e : _this.e1
```
Reads a lot better than repetition down below.

Comment (maxhaton) [reply compiler/src/dmd/expressionsem.d:15223]: Actually I'm a dummy

Comment (dkorpel) [reply compiler/src/dmd/expressionsem.d:15223]: I don't want this function to accept 'nullable' classes, the caller should pass proper instances. Later I'd like to refactor `modifiableLvalue` to so it doesn't accept a `null` parameter either.

**Review by PetarKirov — APPROVED** [review]
LGTM. How feasible is it to write a regression test for this?

**Comment by dkorpel** [comment]
The segfault was triggered by `gccAsmSemantic` calling `modifiableLvalue` with a `null` argument. I could add a test for calling `modifiableLvalue` with a second `null` expression outside an `IN_GCC` block, but I'd rather remove the redundant (public) argument there too.

**Comment by thewilsonator** [comment]
merge conflict

**Review by thewilsonator — APPROVED** [review]

**Comment by ibuclaw** [comment]
> LGTM. How feasible is it to write a regression test for this?

Do we have semantic unittests? Otherwise you could synthesize it in the cxxfrontend test.

**Comment by ibuclaw** [comment]
> > LGTM. How feasible is it to write a regression test for this?
> 
> Do we have semantic unittests? Otherwise you could synthesize it in the cxxfrontend test.

[Yes we do](https://github.com/dlang/dmd/tree/master/compiler/test/unit/semantic).

```patch
diff --git a/compiler/test/unit/semantic/lvalue.d b/compiler/test/unit/semantic/lvalue.d
new file mode 100644
index 0000000000..ab09f154d7
--- /dev/null
+++ b/compiler/test/unit/semantic/lvalue.d
@@ -0,0 +1,33 @@
+// Tests regarding Expression.toLvalue
+//
+// See ../../README.md for information about DMD unit tests.
+
+module semantic.lvalue;
+
+import dmd.dscope : Scope;
+import dmd.expression : TypeExp;
+import dmd.expressionsem : modifiableLvalue;
+import dmd.location : Loc;
+import dmd.globals : global;
+import dmd.mtype : Type;
+
+import support;
+
+@("iasmgcc-lvalue-check")
+unittest
+{
+    /+ We get this when running semantic on:
+        void fun(X...)(X arg)
+        {
+            alias X[0] type;            // an alias to a TypeExp
+            asm { "" : "=m" (type); }   // "=m" requires argument to be modifiable
+        }
+    +/
+    uint errors = global.startGagging();
+
+    scope te = new TypeExp(Loc.initial, Type.tfloat64);
+    te.modifiableLvalue(null, null);
+
+    if (!global.endGagging(errors))
+        assert(false);
+}
```

@dkorpel

**Comment by ibuclaw** [comment]
> The segfault was triggered by `gccAsmSemantic` calling `modifiableLvalue` with a `null` argument. I could add a test for calling `modifiableLvalue` with a second `null` expression outside an `IN_GCC` block, but I'd rather remove the redundant (public) argument there too.

This is the test that triggers the ICE.

https://github.com/gcc-mirror/gcc/blob/b51bfee1beed03872ea0289cb47dd2336d9f528c/gcc/testsuite/gdc.dg/asm1.d#L61

The error is slightly cryptic, so perhaps `gccAsmSemantic` could benefit by not passing in a `null` argument.

The only other place I see that passes `null` to this function is in [DeleteExp](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/expressionsem.d#L8806).

**Review by ibuclaw — CHANGES_REQUESTED** [review]
Add test as posted.

**Comment by dkorpel** [comment]
I removed the nullable parameter from `modifiableLvalue` instead

**Comment by ibuclaw** [comment]
> I removed the nullable parameter from `modifiableLvalue` instead

It looks reasonable to me at least. Will try bootstrapping with it later.

**Pull request merged**